### PR TITLE
Feat/improve kit socket apis

### PR DESCRIPTION
**Pull request by zaripych**

### CODE REVIEW

File: src/setup/background.ts [hunk 1]
```diff
@@ -1,24 +1,29 @@
...
-  if (config.background) {
-    const backgroundConfig = await config.background();
-    const result =
-      config.shouldUseDefaultBackgroundOperations ||
-      typeof config.shouldUseDefaultBackgroundOperations !== 'boolean'
-        ? merge(backgroundConfig, defaultBackground)
-        : backgroundConfig;
+  const useDefault = config.shouldUseDefaultBackgroundOperations ?? true;
+  if (config.background || useDefault) {
+    const backgroundEpics = await (config.background?.() ??
```
Comment (jaycosaur) [inline src/setup/background.ts]: What a beauty!

File: src/shared/auth/epicWithAuth.ts [hunk 2]
```diff
@@ -99,9 +96,7 @@ export function epicWithAuth<E extends ISocketEpicWithAuth<unknown>>(
 ) {
   const authForEpic = Utils.setFunctionName(
     `withAuth.${epic.name}`,
-    (...args: Parameters<E>) => {
-      const [cmd, req, binary, ...rest] = args;
-
+    (cmd: Parameters<E>[0], ctx: Parameters<E>[1]) => {
```
Comment (jaycosaur) [inline src/shared/auth/epicWithAuth.ts]: ```...[cmd, ctx]:Parameters<E>``` Does this work?

File: src/shared/logging/logEpic.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    params.input ?? true,
+    params.logEvents
+  );
+
+  const maybeLogOutgoing = conditionalOperator(
+    params.output ?? true,
+    params.logEvents
+  );
+
+  const fn = (...args: Parameters<E>) => {
```
Comment (jaycosaur) [inline src/shared/logging/logEpic.ts]: ```const fn = (...[commands, ...rest]: Parameters<E>) => {```

What a beauty.

File: src/shared/logging/logger.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    logger !== null &&
+    typeof logger === 'object' &&
+    'withSuffix' in logger &&
+    'parent' in logger
+  );
+}
+
+function taggedLoggerFactory(parent: Logger, suffixes: unknown[] = []) {
+  return (additional: unknown): LoggerWithSuffix => {
+    suffixes.push(additional);
```
Comment (jaycosaur) [inline src/shared/logging/logger.ts]: The devil has arose! And he brings mutations!

```function taggedLoggerFactory(parent: Logger, suffixes: unknown[] = []) {
  const wrappedSuffixes = [...suffixes]
  return (additional: unknown): LoggerWithSuffix => {
    wrappedSuffixes.push(additional);```

Comment (jaycosaur) [reply src/shared/logging/logger.ts]: Unless you genuinly want to mutate the original suffix array...

File: src/shared/logging/logEvents.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      case 'error':
+        params.logger.log(...description, info.error, ...(params.suffix ?? []));
+        break;
+      default:
+        params.logger.log(...description, ...(params.suffix ?? []));
+        break;
+    }
+  } catch (err) {
+    registerError(err);
+    console.error('💥  Something bad happened when logging', err);
```
Comment (jaycosaur) [inline src/shared/logging/logEvents.ts]: lol

File: src/shared/logging/logger.ts (hunk 4 above)
Comment (zaripych) [reply src/shared/logging/logger.ts]: Nah, I somehow got to this state by optimizing number of variables declared

**Pull request merged**